### ISSUE TITLE
Confused about how to pull data correctly/effciently

### ISSUE THREAD

**Issue by Thorbenl** [ZEISS-PiWeb/PiWeb-Api#245]
**I am using:**
[x] C# NuGet
[] REST API

**Used version**
PiWeb Server 8.6

**Question**
For any given Part, I would like to fetch out the measurement values. I am very new to PiWebs infrastructure, so here is a hopefully good outline of our setup:

### Parts

![Screenshot 2024-08-20 at 11 47 52](https://github.com/user-attachments/assets/43c5889f-24bd-46ea-acc9-ca0712739cf5)
(please note that we dont care about M1/M2, so the parent of this, but only whats inside)

And then the values:

![Screenshot 2024-08-20 at 11 48 35](https://github.com/user-attachments/assets/94985344-df23-4d95-8844-2955cbec9c2c)

Then here we can see that screen, and also what i believe we can easily fetch with `dataServiceClient.GetMeasurementValues(partPath)`..

However, what I dont understand, and what we only really need is: 
![Screenshot 2024-08-20 at 11 52 32](https://github.com/user-attachments/assets/0eb80407-430c-4584-8ab6-843d9f743f6c)

So for each measurements, we would liek to extract the measured value fiedl only.

Here is what I tried:

```dotnet
                    var partPath = PathHelper.String2PartPathInformation("APR/PD10003020/3020069/");
                    var parts = await dataServiceClient.GetParts(partPath);
                    var characteristics = await dataServiceClient.GetCharacteristics( partPath );
                    IReadOnlyList<DataMeasurementDto> fetchedMeasurementsWithValues =
                        await dataServiceClient.GetMeasurementValues(partPath,
                            new MeasurementValueFilterAttributesDto
                            {
                                AggregationMeasurements = AggregationMeasurementSelectionDto.All,
                                Deep = true
                            }, cancellationToken: cancellationToken);

                    logger.LogInformation("Polling data from piweb, {number} records ...", fetchedMeasurementsWithValues.Count);
                    foreach (DataMeasurementDto data in fetchedMeasurementsWithValues)
                    {
                        foreach (var dataCharacteristic in data.Characteristics)
                        {
                            logger.LogInformation("data: {data2}", dataCharacteristic.Value.MeasuredValue);
                        }
                    }

```

I am a bit clueless here. What we need is only the measured value for each measurement in each of the parts (so to speak)

Thanks!

**Comment by czthiele** [comment]
Hello @Thorbenl,

at first, your code is correct for what you want to do, you should successfully be fetching the measurements with their values and all attributes.
Your list _fetchedMeasurementsWithValues_ corresponds to your second screenshot, it contains all measurements of the specified part. Inside each measurement there is the dictionary _Characteristics_, containing the measured values for the characteristics (so the things shown in your third screenshot).

The filter class _MeasurementValueFilterAttributesDto_ has some other nice settings you could use. For one this could be the _LimitResult_ option, to limit how many measurements should be returned. This can be useful if the connected server has a really high number of measurements.
Additionally the options _RequestedMeasurementAttributes_ and _RequestedValueAttributes_ are useful to restrict which attributes should be returned. Currently you are getting all attributes, but if you are just interested in a subset you can restrict this using the attribute keys (K-number in the PiWeb configuration), e.g.

`RequestedValueAttributes = new AttributeSelector(new[]{WellKnownKeys.Value.MeasuredValue})` 

would restrict the result so that the value would only contain the measured value (K1). This can save time and data transfer size.

You can find further examples for working with our NuGet here:
https://github.com/ZEISS-PiWeb/PiWeb-Training
https://zeiss-piweb.github.io/PiWeb-Api/sdk/v8.1/
And if you need an overview about the relationships in PiWeb, you can read about this here: https://zeiss-piweb.github.io/PiWeb-Api/general

I hope this gives you a good starting point, feel free to ask if further questions arise.
Best regards
czthiele

**Comment by Thorbenl** [comment]
> AttributeSelector

First of all, thanks a lot of the swift answer, I think we are getting somewhere.

![Screenshot 2024-08-20 at 15 13 23](https://github.com/user-attachments/assets/bcb79570-a244-4e03-aae8-f2ef02782617)

I cant seem to find a way of restricting `characteristics` I am only interested in anything as children:

So I have:
M1
- M1_01
- M1_02
etc.

Im not really interested in the parents (MX) here, just in the children.

Additionally, I am really only interested in the K14 standard value from PiWeb, but I cant really find that reliably.

So again, thanks for helping out. :)

We have done 80 measurements, which i can find on `dataServiceClient.GetMeasurementValues(partPath)` but i really only need to pull out the K14 foreach.

**Comment by czthiele** [comment]
@Thorbenl this is no problem. First of all just as a side note, the parents like M1 are still characteristics, even though they have a similiar icon like parts (this is controlled by some attribute, I think these are calles logical groups). This is why they still appear as characteristics.

You can restrict the measurement search to specific characteristics using the _MeasurementValueFilterAttributesDto_ property _CharacteristicsUuidList_, you simply specify the UUIDs of the characteristics you are interested in. Since you already fetched the list of characteristics, you could filter this list to only contain the correct characteristics (e.g. by name), and give the UUIDs to the filter dto.

At least for the standard config, K14 is a measurement attribute (Part ident), so you would need to specify the requested measurement attributes in the filter like this:
`RequestedMeasurementAttributes = new AttributeSelector(new ushort[]{14})`
Now, in each measurement the _Attributes_ property should only contain the attribute K14 (if there is a value for this).
The entities in PiWeb (like a measurement) also have some extension methods for attributes to easily get them from the _Attributes_ list. You can simply use something like <yourMeasurement>.GetAttribute(14) or .GetAttributeValue(14) to get the attribute K14 if it exists on the entity.

Keep in mind: if K14 is a measured value attribute in your configuration, then you would need to use the _RequestedValueAttributes_ property again, and then they are located one layer deeper, in each characteristics _Attributes_ property.

I hope this brings you one step further.

**Comment by Thorbenl** [comment]
> @Thorbenl this is no problem. First of all just as a side note, the parents like M1 are still characteristics, even though they have a similiar icon like parts (this is controlled by some attribute, I think these are calles logical groups). This is why they still appear as characteristics.
> 
> You can restrict the measurement search to specific characteristics using the _MeasurementValueFilterAttributesDto_ property _CharacteristicsUuidList_, you simply specify the UUIDs of the characteristics you are interested in. Since you already fetched the list of characteristics, you could filter this list to only contain the correct characteristics (e.g. by name), and give the UUIDs to the filter dto.
> 
> At least for the standard config, K14 is a measurement attribute (Part ident), so you would need to specify the requested measurement attributes in the filter like this: `RequestedMeasurementAttributes = new AttributeSelector(new ushort[]{14})` Now, in each measurement the _Attributes_ property should only contain the attribute K14 (if there is a value for this). The entities in PiWeb (like a measurement) also have some extension methods for attributes to easily get them from the _Attributes_ list. You can simply use something like .GetAttribute(14) or .GetAttributeValue(14) to get the attribute K14 if it exists on the entity.
> 
> Keep in mind: if K14 is a measured value attribute in your configuration, then you would need to use the _RequestedValueAttributes_ property again, and then they are located one layer deeper, in each characteristics _Attributes_ property.
> 
> I hope this brings you one step further.

Thanks a lot. This is all I really needed. :) Brilliant.

Last question before we can close this issue, and maybe it is a more conceptual one:

We are using a path similiar to `APR/PDXXXX/XXXX/` where the two route parameters are some identifiers we set up. However, our middle layer is supposed to do polling, because we are waiting for the event based system, as I asked here: #242 .

We don't know these two route parameters, because they technically come from a different system. My question to you is, if we can poll anything new under `APR/*, so we know don't need to introduce more overhead to our stack

Thanks a lot!

**Comment by Thorbenl** [comment]
@czthiele Thanks again :) This makes a lot of sense. And helped a lot. 

I can explain the usecase a bit, without revealing too much: but we really want and need to push the data that we have captured in piweb, into another external system. I am trying to find the smartes way here, and we know polling is our only option now, however, we struggle to find a way of keeping track of what has already been pushed and what not. We really dont want to introduce a mediator database to keep track of that, because thats a lot of infrastructure overhead. 

My last idea here is, to enable something like a CDC table on the piweb database. We own the DB, and we can technically just create something in a different schema, that is just a table that we can pull from our external system.
Even tho we understand a lot now, from your explanation here, which is highly appreciated, we struggle to find a good documentation on how the DB actually pulls out the data, or rather how the REST API operates here.

I cant seem to find any Table called characteristics. What we are aiming for is a trigger a la

`sql
CREATE TRIGGER tr_NewMeasurementMade
    ON ....
    AFTER INSERT, UPDATE
                      AS
BEGIN
BEGIN TRY
        DECLARE @RowId BIGINT
        ...
        -- Select values into variables
SELECT @RowId = Id,
       ...
FROM INSERTED

        ....
        END

    END TRY
    BEGIN CATCH
            DECLARE @ErrorMessage NVARCHAR(MAX) = ERROR_MESSAGE();
            -- Insert the error message into the log table
    INSERT INTO NewSampleTriggerErrorLog (ErrorMessage) VALUES (@ErrorMessage);
    -- Print the error message to the console if needed
    PRINT 'Error: ' + @ErrorMessage;
    END CATCH
END
```

Ideally, and we think it would combine our findings here, if the new table or view, could only put the minimum necesary information.
Basically 
![image](https://github.com/user-attachments/assets/83b463a6-1a35-4099-9d85-1ee718034013)
this in form of a database table. so we can pick the 3-4 fields we want from here

**Comment by czthiele** [comment]
Hi @Thorbenl,

I see that you put a lot of thought into ways to solve this, but unfortunately the database approach is no real option. The PiWeb Server will not start if database triggers are present for its database. This has a long history of customers trying to directly work with the database, failing, and getting all sorts of weird behaviours and "bugs" in PiWeb, which have cost a lot of time analyzing.
Additionally, our database schema (the way the tables are working together, not the "schema" like [dbo]) can change between releases, thus breaking your trigger (or an error in the trigger breaking PiWeb).
This is why we always state to use the REST API or our C# NuGet, since this has versioning and we pay a lot of attention to compatibility. A lot of new features are added in a way that does not break the REST Api, hence the fact that we are still on major interface version 1.X.

I do understand that you do not want the overhead of keeping track of already synchronized measurements, but I think that there is no way around this for now, you still need to keep at least the UUIDs stored somewhere, unless your destination system also supports some kind of data retrieval. In this case you could use the measurement UUID as a key to check if this measurement is already present in your destination system.
You can also restrict the request to PiWeb in a way that only measurements in a specific time range are returned, minimizing the amount of checks for the "already synchronized" case. See [here](https://zeiss-piweb.github.io/PiWeb-Api/sdk/v8.1/#-example--fetching-measurements-in-a-time-range) for an example.

I know that this is not exactly what you want, but with high probability the only option until the coming release.
Best regards,
czthiele

**Comment by Thorbenl** [comment]
> Hi @Thorbenl,
> 
> I see that you put a lot of thought into ways to solve this, but unfortunately the database approach is no real option. The PiWeb Server will not start if database triggers are present for its database. This has a long history of customers trying to directly work with the database, failing, and getting all sorts of weird behaviours and "bugs" in PiWeb, which have cost a lot of time analyzing. Additionally, our database schema (the way the tables are working together, not the "schema" like [dbo]) can change between releases, thus breaking your trigger (or an error in the trigger breaking PiWeb). This is why we always state to use the REST API or our C# NuGet, since this has versioning and we pay a lot of attention to compatibility. A lot of new features are added in a way that does not break the REST Api, hence the fact that we are still on major interface version 1.X.
> 
> I do understand that you do not want the overhead of keeping track of already synchronized measurements, but I think that there is no way around this for now, you still need to keep at least the UUIDs stored somewhere, unless your destination system also supports some kind of data retrieval. In this case you could use the measurement UUID as a key to check if this measurement is already present in your destination system. You can also restrict the request to PiWeb in a way that only measurements in a specific time range are returned, minimizing the amount of checks for the "already synchronized" case. See [here](https://zeiss-piweb.github.io/PiWeb-Api/sdk/v8.1/#-example--fetching-measurements-in-a-time-range) for an example.
> 
> I know that this is not exactly what you want, but with high probability the only option until the coming release. Best regards, czthiele

Thanks :)

How are we looking with the event-based streaming? Are you on track for an october release? :D

**Comment by Thorbenl** [comment]
@czthiele Could you give an indication here on if Event Streaming is going to make it into the next release?

**Comment by czthiele** [comment]
Hi @Thorbenl,

sorry for the late reply, I was absent last week and GitHub did not notify my colleagues due to the personal @ mention.
Of course, the event feature will be part of the next release, as promised. We are currently working on the last finishing touches, e.g. the documentation.

Best regards
czthiele

**Comment by czthiele** [comment]
Hello @Thorbenl,

good news! The release 2025.R1 will be available soon, including the new events feature.
You can find more information in the endpoint documentation here [DataServiceRest Events](https://zeiss-piweb.github.io/PiWeb-Api/dataservice/v1.11/#ds-events), and also here [RawDataServiceRest Events](https://zeiss-piweb.github.io/PiWeb-Api/rawdataservice/v1.8/#rs-events).

A C# ready-to-run sample application can be found as part of the [API Sample training application here](https://github.com/ZEISS-PiWeb/PiWeb-Training?tab=readme-ov-file#piwebapi---events).
I hope this will help you get started, and as always feel free to ask if questions arise.

Best regards,
czthiele

**Comment by Thorbenl** [comment]
@czthiele  Thanks for the info! Do you have a release date? Or is it already out?

**Comment by czthiele** [comment]
Hi @Thorbenl,

PiWeb 2025.R1 is released and available in the [ZEISS Portal](https://portal.zeiss.com/download-center/softwares/imt).
You can use the newest [API NuGet v9.0.0](https://www.nuget.org/packages/Zeiss.PiWeb.Api.Rest/9.0.0), but you are not forced to use this version. 
Please note, the events endpoint is part of the PiWeb Server 2025.R1, its usage is not (yet) directly built into the API NuGet. For C# you would use the Microsoft.AspNetCore.SignalR.Client NuGet to connect to this endpoint, please take a look at the [PiWeb.API - Events Training Sample](https://github.com/ZEISS-PiWeb/PiWeb-Training?tab=readme-ov-file#piwebapi---events). This will show you how to utilize the new endpoint and you can take the C# sample as a starting point.

We may decide to incorporate the functionality into the main NuGet in the future, but for the first step this release we have the public endpoint with documentation.

Best regards,
czthiele

**Comment by Thorbenl** [comment]
> Hi @Thorbenl,
> 
> PiWeb 2025.R1 is released and available in the [ZEISS Portal](https://portal.zeiss.com/download-center/softwares/imt). You can use the newest [API NuGet v9.0.0](https://www.nuget.org/packages/Zeiss.PiWeb.Api.Rest/9.0.0), but you are not forced to use this version. Please note, the events endpoint is part of the PiWeb Server 2025.R1, its usage is not (yet) directly built into the API NuGet. For C# you would use the Microsoft.AspNetCore.SignalR.Client NuGet to connect to this endpoint, please take a look at the [PiWeb.API - Events Training Sample](https://github.com/ZEISS-PiWeb/PiWeb-Training?tab=readme-ov-file#piwebapi---events). This will show you how to utilize the new endpoint and you can take the C# sample as a starting point.
> 
> We may decide to incorporate the functionality into the main NuGet in the future, but for the first step this release we have the public endpoint with documentation.
> 
> Best regards, czthiele

Awesome work. Thanks a look. I appreciate you guys taking the time. 

I think its time to clue this issue also, and if anything else arises, I will reach out via different channels, or a new issue here :)
